cip/CIP2015-05-13-EXISTS.adoc

```diff
@@ -0,0 +1,174 @@
+
+----
+expression = …
+           | property exists
+           | subquery exists
+           | simple subquery exists
+            ;
+
+property exists = "EXISTS", "(", expression ")"
+                ;
+
+subquery exists = "EXISTS", "{", read only clause, { read only clause }, "}" ;
+
+simple subquery exists = "EXISTS", "{", simple match, "}" ;
+
+simple match = pattern, { ",", pattern }, [ "WHERE", predicate ] ;
+
+----
+Note: the read-only clause above includes WITH but not RETURN.
+
+=== Semantics
+
+All forms of +EXIST(...)+ accomplish the same task - check if a particular pattern exists in the database.
+They are expressions, and as such must be side-effect free; that is, the subqueries in +EXIST+ must not be updating queries.
+Regarding nomenclature, all forms of +EXISTS+ are scalar expressions - they return a single boolean value, and this value is never +null+.
+
+==== Property exists
+
+In the property checking form, any expression is accepted as the input.
+If the input expression returns a +null+ value, +EXISTS(...)+ returns false, and for all other values, it will return +true+.
+This form of +EXISTS(...)+ is equivalent to checking for +null+ using +IS NOT NULL+; i.e. +EXISTS(expr)+ is equivalent to +expr IS NOT NULL+.
+
+This form is there to make property existence checking idiomatic in Cypher - we do not want the user thinking of missing properties as checking for +null+ values.
+
+==== Subquery exists
+When checking for the existence of subqueries, variables and parameters from the outer query are available in the subquery.
+Variables introduced in the subquery are not available on the outside.
+
+For each matching subgraph evaluated with +EXISTS{...}+, the result value must be true if the subquery finds at least one matching row.
+If no matches are found, false should be returned.
+
+The subquery used in +EXISTS{...}+ has to follow the following rules:
+
+ * It has to be a read-only query. Updating the graph as part of a predicate is not allowed.
+ * It cannot end in a +RETURN+ clause, nor may it contain +UNION+
+ * If +WITH+ ends the subquery, it should be able to change the cardinality of the subquery in an for +EXISTS(...)+ interesting way.
+In other words, they should include +WHERE+ or +SKIP+, since they could turn a query that returns rows into one that does not.
+
+==== Simple subquery exists
+When the subquery can be described with a single +MATCH-WHERE+ clause, the +MATCH+ keyword can be omitted.
+The difference between this form and a simple pattern predicate, which is already available is that this form allows for introducing new variables inside the +EXISTS(...)+ scope.
+
+=== Examples
+_Example 1A:_
+
+Return all nodes that have a property named slogan.
+[source, cypher]
+----
+MATCH (actor)
+WHERE EXISTS(actor.slogan)
+RETURN actor
+----
+
+_Example 1B:_
+Return all nodes with both first name and second name:
+[source, cypher]
+----
+MATCH (actor)
+WHERE EXISTS(actor.firstName + actor.lastName)
+RETURN actor
```

Comment:
Wow, this is an interesting side effect of the semantics of `null`.

But in this case I guess it would only work if the two properties are of compatible types. In this example they are likely both `string?`, but what if one was `string?` and the other `number?`?

Oh wait... Cypher does disgusting type coercion to string when one of the operands to `+` is a string... :/